File: 0x10 基本数据结构/0x12 队列/AcWing132.cpp
#include <bits/stdc++.h>
using namespace std;
const int N = 1010, M = 1e6 + 10;
int teamid[M]; // 存储编号为 x 的人属于哪一个队伍（队伍编号 0 ~ n-1）

int main() {
  int n, C = 1;
  while (cin >> n, n) {
    queue<int> team;
    queue<int> person[N];
    printf("Scenario #%d\n", C ++ );
    for (int i = 0; i < n; i ++ ) {
      int cnt;
      cin >> cnt;
      while (cnt -- ) {
        int x;
        cin >> x;
        teamid[x] = i;
      }
    }

    string command;
    while (cin >> command, command != "STOP")
      if (command == "ENQUEUE") {
        int x;
        cin >> x;
        int tid = teamid[x];
        if (person[tid].empty()) team.push(tid);
        person[tid].push(x);
      }
      else {
        int tid = team.front();
        auto &q = person[tid];
        cout << q.front() << endl;
        q.pop();
        if (q.empty()) team.pop();
      }

    cout << endl;
  }

  return 0;
}

File: 0x10 基本数据结构/0x12 队列/AcWing132.py
import sys
from collections import deque

def main():
    input_data = sys.stdin.read().split()
    if not input_data:
        return
        
    iterator = iter(input_data)
    scenario = 1
    
    while True:
        try:
            n = int(next(iterator))
        except StopIteration:
            break
            
        if n == 0:
            break
            
        print(f"Scenario #{scenario}")
        scenario += 1
        
        team_id = {}
        for i in range(n):
            cnt = int(next(iterator))
            for _ in range(cnt):
                team_id[next(iterator)] = i
                
        team_queue = deque()
        person_queues = [deque() for _ in range(n)]
        
        while True:
            cmd = next(iterator)
            if cmd == "STOP":
                break
            elif cmd == "ENQUEUE":
                x = next(iterator)
                tid = team_id[x]
                if not person_queues[tid]:
                    team_queue.append(tid)
                person_queues[tid].append(x)
            else:
                tid = team_queue[0]
                print(person_queues[tid].popleft())
                if not person_queues[tid]:
                    team_queue.popleft()
                    
        print()

if __name__ == '__main__':
    main()